netket/experimental/operator/_fermions_2nd.py

```diff
@@ -0,0 +1,695 @@
+
+                    new_term = True
+
+            if pad:
+                n_c = (b + 1) * max_conn
+
+            sections[b] = n_c
+
+        if pad:
+            return x_prime, mels
+        else:
+            return x_prime[:n_c], mels[:n_c]
+
+    def _op__matmul__(self, other):
+        if not isinstance(other, FermionOperator2nd):
+            return NotImplementedError
+        if not self.hilbert == other.hilbert:
+            raise ValueError(
+                f"Can only multiply identical hilbert spaces (got A@B, A={self.hilbert}, B={other.hilbert})"
+            )
+        terms = []
+        weights = []
+        for t, w in self._operators.items():
+            for to, wo in other._operators.items():
+                terms.append(tuple(t) + tuple(to))
+                weights.append(w * wo)
+        if other._constant != 0.0:
+            for t, w in self._operators.items():
+                terms.append(tuple(t))
+                weights.append(w * other._constant)
+        if self._constant != 0:
+            for t, w in other._operators.items():
+                terms.append(tuple(t))
+                weights.append(w * self._constant)
+        constant = self._constant * other._constant
+        dtype = np.promote_types(self.dtype, other.dtype)
+        return FermionOperator2nd(
+            self.hilbert, terms, weights=weights, constant=constant, dtype=dtype
+        )
+
+    def __radd__(self, other):
+        return self + other
+
+    def __add__(self, other):
+        op = self.copy()
+        op += other
+        return op
+
+    def __iadd__(self, other):
+        if np.issubdtype(type(other), np.number):
+            if other != 0.0:
+                self._constant += other
+            return self
+        if not isinstance(other, FermionOperator2nd):
+            raise NotImplementedError
+        if not self.hilbert == other.hilbert:
+            raise ValueError(
+                f"Can only add identical hilbert spaces (got A+B, A={self.hilbert}, B={other.hilbert})"
+            )
+
+        for t, w in other._operators.items():
+            if t in self._operators.keys():
+                self._operators[t] += w
+            else:
+                self._operators[t] = w
+        self._constant += other._constant
+        self._dtype = np.promote_types(self.dtype, other.dtype)
+        return self
+
+    def __isub__(self, other):
+        self += -other
+        return self
+
+    def __sub__(self, other):
+        return self + (-other)
+
+    def __rsub__(self, other):
+        return other + (-self)
+
+    def __neg__(self):
+        return -1 * self
+
+    def __rmul__(self, scalar):
+        return self * scalar
+
+    def __imul__(self, scalar):
+        if not np.issubdtype(type(scalar), np.number):
+            # we will overload this as matrix multiplication
+            raise Exception(
+                "cannot inplace multiply two operators, since it changes the internal structure"
+            )
```

Comment:
you must check that the dtype of internal structures does not change.
```suggestion
            )
        if not np.can_cast(type(other), self.dtype, casting="same_kind"):
             raise ValueError(
                 f"Cannot add inplace operator with dtype {type(other)} "
                 f"to operator with dtype {self.dtype}"
             )
```